src/data/finsum.lean

```diff
@@ -0,0 +1,724 @@
+than `finset.sum` when issues arise with `finset` and `fintype` being data.
+
+## Main definitions
+
+Variables:`
+
+`(α : Type u)`
+`(s : set α)`
+`(M : Type v) [add_comm_monoid M]`
+`(f : α → M)`
+
+Definitions in this file:
+
+* `univ' α : finset α` : `α` as a finset if `α` is finite, and the empty finset otherwise
+
+* `finsum f : M` : the sum of `f x` as `x` ranges over the support of `f`, if it's finite.
+   Zero otherwise.
+
+* `finsum_in s f : M` : the sum of `f x` as `x` ranges over the elements of `s` for which `f x ≠ 0`,
+  if this is a finite sum. Zero otherwise.
+
+## Notation
+
+* `∑ᶠ i, f i` and `∑ᶠ i : α, f i` for `finsum f`
+
+* `∑ᶠ i in s, f i` for `finsum_in s f`
+
+## Implementation notes
+
+There is no `finprod` right now. This is because `function.support` and `finsupp` etc
+are all focussed on 0, and there are no analogous definitions for 1.
+
+## Todo
+
+We did not add `is_finite (X : Type) : Prop`, because it is simply `nonempty (fintype X)`.
+There is work on `fincard` in the pipeline, which returns the cardinality of `X` if it
+is finite, and 0 otherwise.
+
+## Tags
+
+finsum, finsum_in, univ', finiteness
+-/
+
+open_locale classical
+
+universes u v w
+
+-- TODO move to data.finset.something?
+/-- `set.to_finset' s` is the finset corresponding to `s` if `s` is finite, and the empty set
+    otherwise -/
+noncomputable def set.to_finset' {α : Type u} (s : set α) : finset α :=
+if h : s.finite then h.to_finset else ∅
+
+theorem set.to_finset'_eq_to_finset {α : Type u} {s : set α} [fintype s] :
+  s.to_finset' = s.to_finset :=
+begin
+  have h : s.finite := ⟨infer_instance⟩,
+  unfold set.to_finset',
+  rw dif_pos h,
+  unfold set.finite.to_finset,
+  congr',
+end
+
+/-!
+
+# finset.univ' -- noncomputable univ (empty set if infinite)
+
+-/
+
+namespace finset
+
+/-- `univ' α` is the finset corresponding to all of α if α is finite, and the empty set otherwise.
+Note that it is noncomputable -/
+noncomputable def univ' (α : Type u) : finset α :=
+if h : nonempty (fintype α) then (classical.choice h).elems else ∅
+
+variable {α : Type u}
+
+/-- `univ' α` is `finset.univ` if `α` is a fintype. -/
+lemma univ'_eq_univ [h : fintype α] : univ' α = univ :=
+by convert (dif_pos (nonempty.intro h))
+
+lemma univ'_eq_empty (h : ¬ nonempty (fintype α)) : univ' α = ∅ :=
+dif_neg h
+
+lemma univ'_eq_empty' (h : fintype α → false) : univ' α = ∅ :=
+univ'_eq_empty $ not_nonempty_fintype.2 $ ⟨h⟩
+
+@[simp] lemma mem_univ' [h : fintype α] (x : α) : x ∈ univ' α :=
+(@univ'_eq_univ _ h).symm ▸ mem_univ _
+
+@[simp] lemma mem_univ'' [h : nonempty (fintype α)] (x : α) : x ∈ univ' α :=
+@mem_univ' α (nonempty.some h) x
+
+end finset
+
+/-! # noncomputable finite sums -/
+
+section finsum
+
+variables {α : Type u} {M : Type v} [add_comm_monoid M]
+
+/-- Sum of `f x` as `x` ranges over the elements of the support of `f`, if it's finite. Zero
+  otherwise. -/
+noncomputable def finsum (f : α → M) : M :=
+(function.support f).to_finset'.sum f
```

Comment:
If you replace `support` by its definition, then you could make this the `to_additive` of the multiplicative version, right?
Is it worth it to avoid `support` just so that we can also have `finprod` (at least for the basic API)?

Reply:
If we replace support with something similar to `{x | f x \neq 1}` then we can simply add a `to_additive` in front and from the small experiment I've done so far, everything remains to work. This is a pretty big refactor however so I'm not sure if its a good idea after all.

Reply:
This suggestion makes us lose all the `support` API. Isn't the correct thing to do to make `function.mulsupport`, the stuff which doesn't get mapped to 1? That doesn't look too bad, to be honest. But then we have to refactor all the proofs or get Yury to write `@[to_multiplicative]` ;-) I've added it as a TODO in the module docstring.